doc/dev/adr/0020-large-message-chunking.md

```diff
@@ -0,0 +1,100 @@
+﻿# ADR 20: Large Message Chunking in MQTT Protocol
+
+## Status
+Proposed
+
+## Context
+The MQTT protocol has inherent message size limitations imposed by brokers and network constraints. Azure IoT Operations scenarios often require transmitting payloads that exceed these limits (e.g., firmware updates, large telemetry batches, complex configurations). Without a standardized chunking mechanism, applications must implement their own fragmentation strategies, leading to inconsistent implementations and interoperability issues.
+
+## Decision
+We will implement sdk-level message chunking as part of the Protocol layer to transparently handle messages exceeding the MQTT broker's maximum packet size.
+
+**The chunking mechanism will**:
+- Be enabled/disabled by a configuration setting.
+- Use standardized user properties for chunk metadata:
+   - The `__chunk` user property will contain a JSON object with chunking metadata.
+   - The JSON structure will include:
+     ```json
+     {
+       "messageId": "unique-id-for-chunked-message",
+       "chunkIndex": 0,
+       "timeout" : "00:00:10",
+       "totalChunks": 5,
+       "checksum": "message-hash"
+     }
+     ```
+   - `messageId, chunkIndex, timeout` - present for every chunk; `totalChunks, checksum` - present only for the first chunk.
+
+**Chunk size calculation**:
+- Maximum chunk size will be derived from the MQTT CONNECT packet's Maximum Packet Size.
+- A static overhead value will be subtracted from the Maximum Packet Size to account for MQTT packet headers, topic name, user properties, and other metadata.
```

Comment:
1) I don't like this as a configuration option.
2) We can't guarantee correctness even with the option, and I don't think we should lock out sufficiently large user properties by design.
3) Header size could fluctuate, so a static value seems insufficient.

While difficult, it would seem to me that we should attempt to do it dynamically. A static value to start with might be a good first step, but I don't like baking it into the config options and setting it for the life of the envoy.

There might also be an argument for handling the edge case of header offset entirely reactively in the name of only handling this (unlikely) case lazily in order to reduce computation overhead in the general case (assuming there is client-side validation).